Enable riscv64 cross build

crazy-max opened the pull request:
needs #50862

follow-up
* #43529
* https://github.com/moby/moby/pull/43553#issuecomment-1115315857
* https://github.com/moby/moby/pull/46028

fixes #44319

**- What I did**

Add support for `riscv64` cross build.

**- How I did it**

**- How to verify it**

```
$ docker buildx bake all --set *.platform=linux/riscv64
$ file bundles/binary/*
bundles/binary/containerd:                    ELF 64-bit LSB executable, UCB RISC-V, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=067d136378178bc72ecb590040cdf4190b47008f, stripped
bundles/binary/containerd-shim-runc-v2:       ELF 64-bit LSB executable, UCB RISC-V, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=feb6112258600b606528ee2d301dd75d5256b327, stripped
bundles/binary/ctr:                           ELF 64-bit LSB executable, UCB RISC-V, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=42245b90973389524046bf8d0334924738c8ff2a, stripped   
bundles/binary/docker-init:                   ELF 64-bit LSB executable, UCB RISC-V, RVC, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=811170b358ecfdb3e0ec313c4944766a353a1d7a, for GNU/Linux 4.15.0, stripped
bundles/binary/docker-proxy:                  ELF 64-bit LSB executable, UCB RISC-V, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=11a9512aba7e3704185be98889aa2fa4f16c6295, not stripped
bundles/binary/dockerd:                       ELF 64-bit LSB executable, UCB RISC-V, RVC, double-float ABI, version 1 (GNU/Linux), statically linked, BuildID[sha1]=bcb3af8af199e49b330784486a7ced3e5ad09363, for GNU/Linux 4.15.0, with debug_info, not stripped
bundles/binary/dockerd-rootless-setuptool.sh: POSIX shell script, ASCII text executable
bundles/binary/dockerd-rootless.sh:           POSIX shell script, Unicode text, UTF-8 text executable
bundles/binary/rootlesskit:                   ELF 64-bit LSB executable, UCB RISC-V, double-float ABI, version 1 (SYSV), statically linked, BuildID[sha1]=a210c22b60ee086017e2607a9c16fb46ae20144c, with debug_info, not stripped
bundles/binary/runc:                          ELF 64-bit LSB executable, UCB RISC-V, RVC, double-float ABI, version 1 (GNU/Linux), statically linked, BuildID[sha1]=95daf11e3014ee1b459d59cf92b9728967c0ec4e, for GNU/Linux 4.15.0, with debug_info, not stripped
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

cc @AkihiroSuda

crazy-max committed:
Enable riscv64 cross build

Signed-off-by: CrazyMax <1951866+crazy-max@users.noreply.github.com>

AkihiroSuda commented:
> bundles/binary/runc

Needs to be built from the main branch ( https://github.com/opencontainers/runc/pull/3446 )

crazy-max commented:
> > bundles/binary/runc
> 
> Needs to be built from the main branch ( [opencontainers/runc#3446](https://github.com/opencontainers/runc/pull/3446) )

Ah didn't catch it was not backported to 1.1 branch. Guess it will be for next 1.2?

crazy-max commented on Dockerfile:
```diff
@@ -281,7 +316,17 @@ RUN git init . && git remote add origin "https://github.com/opencontainers/runc.
 # consider updating runc in vendor.mod accordingly.
 ARG RUNC_VERSION=v1.1.7
-RUN git fetch -q --depth 1 origin "${RUNC_VERSION}" +refs/tags/*:refs/tags/* && git checkout -q FETCH_HEAD
+ARG TARGETARCH
+RUN <<EOT
+  set -e
+  if [ "$TARGETARCH" = "riscv64" ]; then
+    # riscv64 is not supported by latest runc 1.1.7, so use the release 1.1
+    # branch that brings support for riscv64 (https://github.com/opencontainers/runc/pull/3905).
+    RUNC_VERSION=release-1.1
+  fi
+  git fetch -q --depth 1 origin "${RUNC_VERSION}" +refs/tags/*:refs/tags/*
+  git checkout -q FETCH_HEAD
+EOT
```
@AkihiroSuda Made this change while waiting for next patch release (related to https://github.com/opencontainers/runc/pull/3905)

AkihiroSuda replied:
v1.1.8 is now available https://github.com/opencontainers/runc/releases/tag/v1.1.8

neersighted commented on Dockerfile:
```diff
@@ -1,8 +1,12 @@
+# be able to build against armel arch
+# TODO: remove ubuntu base when riscv64 is supported by golang debian base image
+ARG DEBIAN_BASE="debian:bullseye"
+ARG UBUNTU_BASE="ubuntu:22.04"
```
I'd like to block this new target on Debian support instead of introducing Ubuntu.

crazy-max replied:
Unfortunately debian doesn't provide any cross pkg for riscv64 architecture (even sid). Also the ubuntu image is just used for the riscv64 arch.

neersighted replied:
https://lwn.net/Articles/939095/

Good news :)

crazy-max replied:
Just in time!

thaJeztah commented:
```
3.803 # runtime/cgo
3.803 In file included from _cgo_export.c:3:
3.803 /usr/include/stdlib.h:25:10: fatal error: 'bits/libc-header-start.h' file not found
54.17 + rm -f /go/src/github.com/docker/docker/go.mod
```


Also seeing a failure when compiling tini; should we also contribute changes there to add riscv64 to their matrix? (there's some patches not released yet https://github.com/krallin/tini/compare/v0.19.0...master)

```
 > [tini-build 4/4] RUN --mount=from=tini-src,src=/usr/src/tini,rw     --mount=type=cache,target=/root/.cache/go-build,id=tini-build-linux/riscv64 <<EOT (set -e...):
0.182   Could not find compiler set in environment variable CC:
0.182 
0.182   riscv64-linux-gnu-gcc.
0.182 Call Stack (most recent call first):
0.182   CMakeLists.txt:2 (project)
0.182 
0.182 
0.182 CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
```

crazy-max commented:
> Also seeing a failure when compiling tini; should we also contribute changes there to add riscv64 to their matrix? (there's some patches not released yet [krallin/tini@v0.19.0...master](https://github.com/krallin/tini/compare/v0.19.0...master))

No that's an issue with xx where we skip installing riscv64 cross pkgs if not sid:

```
#41 0.155 skipping packages installation on riscv64. riscv64 currently only supported on sid
```

Also needs to fix sources list location in xx:

```
#41 0.155 grep: /etc/apt/sources.list: No such file or directory
```

thaJeztah commented:
@crazy-max I see you just rebased; FYI: this fix was just merged;

- https://github.com/moby/moby/pull/48455

crazy-max commented:
> with xx where we skip installing riscv64 cross pkgs if not sid:
> 
> ```
> #41 0.155 skipping packages installation on riscv64. riscv64 currently only supported on sid
> ```

Still needs work to fix cross pkgs on xx: https://github.com/moby/moby/actions/runs/10791920480/job/29930441711#step:5:708

```
#44 [tini-build 3/4] RUN --mount=type=cache,sharing=locked,id=moby-tini-aptlib,target=/var/lib/apt     --mount=type=cache,sharing=locked,id=moby-tini-aptcache,target=/var/cache/apt         xx-apt-get install -y --no-install-recommends             gcc             libc6-dev             pkg-config
#44 0.143 skipping packages installation on riscv64. riscv64 currently only supported on sid
#44 DONE 0.2s

#45 [tini-build 4/4] RUN --mount=from=tini-src,src=/usr/src/tini,rw     --mount=type=cache,target=/root/.cache/go-build,id=tini-build-linux/riscv64 <<EOT (set -e...)
#45 0.149 CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
#45 0.149   Compatibility with CMake < 2.8.12 will be removed from a future version of
#45 0.149   CMake.
#45 0.149 
#45 0.149   Update the VERSION argument <min> value or use a ...<max> suffix to tell
#45 0.149   CMake that the project does not need compatibility with older versions.
#45 0.149 
#45 0.149 
#45 0.169 CMake Error at /usr/share/cmake-3.25/Modules/CMakeDetermineCCompiler.cmake:49 (message):
#45 0.169   Could not find compiler set in environment variable CC:
#45 0.169 
#45 0.169   riscv64-linux-gnu-gcc.
#45 0.169 Call Stack (most recent call first):
#45 0.169   CMakeLists.txt:2 (project)
#45 0.169 
#45 0.176 
#45 0.176 CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
#45 0.178 -- Configuring incomplete, errors occurred!
#45 0.178 See also "/go/src/github.com/krallin/tini/CMakeFiles/CMakeOutput.log".
#45 ERROR: process "/bin/sh -c   set -e\n  CC=$(xx-info)-gcc cmake .\n  make tini-static\n  xx-verify --static tini-static\n  mkdir /build\n  mv tini-static /build/docker-init\n" did not complete successfully: exit code: 1
```

thaJeztah committed:
Dockerfile update to debian 13 "trixie"

Update our Dockerfile to use debian 13 as base. Some notes about architectures
from the [announcement][1];

> This release for the first time officially supports the riscv64 architecture,
> allowing users to run Debian on 64-bit RISC-V hardware and benefit from all
> Debian 13 features. A total of seven architectures are officially supported
> for trixie:
>
> - 64-bit PC (amd64),
> - 64-bit ARM (arm64),
> - ARM EABI (armel),
> - ARMv7 (EABI hard-float ABI, armhf),
> - 64-bit little-endian PowerPC (ppc64el),
> - 64-bit little-endian RISC-V (riscv64),
> - IBM System z (s390x)
>
> i386 is no longer supported as a regular architecture: there is no official
> kernel and no Debian installer for i386 systems. The i386 architecture is
> now only intended to be used on a 64-bit (amd64) CPU. Users running i386
> systems should not upgrade to trixie. Instead, Debian recommends either
> reinstalling them as amd64, where possible, or retiring the hardware.
>
> trixie will be the last release for the armel architecture. See
> [5.1.3. Last release for armel][2] in the release notes for more information
> on our ARM EABI support.

[1]: https://www.debian.org/News/2025/20250809
[2]: https://www.debian.org/releases/trixie/release-notes/issues.html#last-release-for-armel

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Dockerfile: install e2fsprogs to fix missing mkfs.ext4

It's used in at least one test;

    === FAIL: amd64.integration-cli TestDockerDaemonSuite/TestDaemonNoSpaceLeftOnDeviceError (0.11s)
        docker_cli_daemon_test.go:1065: assertion failed:
            Command:  mkfs.ext4 -F /tmp/no-space-left-on-device-test183525913/testfs.img
            ExitCode: 127
            Error:    exec: "mkfs.ext4": executable file not found in $PATH
            Stdout:
            Stderr:

            Failures:
            ExitCode was 127 expected 0
            Expected no error
        --- FAIL: TestDockerDaemonSuite/TestDaemonNoSpaceLeftOnDeviceError (0.11s)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

crazy-max commented:
@thaJeztah @akerouanton Looks good with xx 1.7.0 on trixie but seems trixie breaks some network related tests?

thaJeztah commented:
Yeah; @akerouanton was still looking into the networking on Trixie; there's an incompatibility with `iptables` on that version

thaJeztah commented:
more details on this one as well;

- https://github.com/moby/moby/pull/50726
